src/Swarm/Game/World.hs

```diff
@@ -247,15 +247,27 @@ lookupEntityM c = do
   lookupEntity c <$> get @(World t e)
 
 -- | Update the entity (or absence thereof) at a certain location,
---   returning an updated 'World'.  See also 'updateM'.
-update :: Coords -> (Maybe e -> Maybe e) -> World t e -> World t e
+--   returning an updated 'World' and a Boolean indicating whether.
+--   the update changed the entity here.
+--   See also 'updateM'.
+update :: Eq e => Coords -> (Maybe e -> Maybe e) -> World t e -> (World t e, Bool)
 update i g w@(World f t m) =
-  World f t (M.insert i (g (lookupEntity i w)) m)
+  (wNew, entityAfter /= entityBefore)
```

Comment:
I suppose this equality check may be expensive, since `Entity` is a wide record.  Maybe if so long as the result of the equality check is consumed lazily, it is ok for the comparison code to be here?

Reply:
You could also compare their hashes instead of checking them for equality directly, i.e.  ``(/=) `on` view entityHash``.